R/Data-class.R

```diff
@@ -30,147 +30,117 @@ NULL
-        ## if two subsequent cohorts are at the same dose.
-        cohort <- as.integer(c(1, 1 + cumsum(diff(x) != 0)))
-    }
-
-    ## then initialize the Data object
-    ## (in this case just putting arguments into slots)
-    ret <- .Data(x=as.numeric(x),
-                 y=safeInteger(y),
-                 ID=safeInteger(ID),
-                 cohort=safeInteger(cohort),
-                 doseGrid=doseGrid,
-                 nObs=length(x),
-                 nGrid=length(doseGrid),
-                 xLevel=matchTolerance(x=x, table=doseGrid),
-                 placebo=placebo)
-    return(ret)
-}
-validObject(Data())
+#' `Data`
+#'
+#' @description `r lifecycle::badge("stable")`
+#'
+#' This is the function for initializing a [`Data-class`] object.
+#'
+#' Note that \code{ID} and \code{cohort} can be missing, then a warning
+#' will be issued and the variables will be filled with default
+#' IDs and best guesses, respectively.
+#'
+#' @param x (`numeric`)\cr the doses for the patients.
+#' @param y (`integer`)\cr the vector of toxicity events (0 or 1 integers).
+#' You can also supply numeric vectors, but these will then be converted to integers.
+#' @param ID (`integer`)\cr unique patient IDs (integer vector).
+#' @param cohort (`integer`)\cr the cohort indices (sorted values from \{0, 1, 2, ...\}).
+#' @param doseGrid (`numeric`)\cr the vector of all possible doses.
+#' @param placebo (`logical`)\cr if `TRUE` the first dose level
+#' in the `doseGrid` is considered as PLACEBO.
+#' @param \dots not used.
+#' @return the initialized [`Data-class`] object.
+#'
+#' @export
+#' @keywords programming
+Data <- function(x = numeric(),
+                 y = integer(),
+                 ID = integer(),
+                 cohort = integer(),
+                 doseGrid = numeric(),
+                 placebo = FALSE,
+                 ...) {
+  ## sort the dose grid
+  doseGrid <- as.numeric(sort(unique(doseGrid)))
+
+  ## check IDs
+  if (length(ID) == 0 && length(x) > 0) {
+    warning("Used default patient IDs!")
+    ID <- seq_along(x)
+  }
 
+  ## check cohort indices
+  if (length(cohort) == 0 && length(x) > 0) {
+    warning("Used best guess cohort indices!")
+    ## This is just assuming that consecutive patients
+    ## in the data set are in the same cohort if they
+    ## have the same dose. Note that this could be wrong,
+    ## if two subsequent cohorts are at the same dose.
+    cohort <- as.integer(c(1, 1 + cumsum(diff(x) != 0)))
+  }
 
-## ============================================================
+  ## then initialize the Data object
+  ## (in this case just putting arguments into slots)
+  ret <- .Data(
+    x = as.numeric(x),
+    y = safeInteger(y),
+    ID = safeInteger(ID),
+    cohort = safeInteger(cohort),
+    doseGrid = doseGrid,
+    nObs = length(x),
+    nGrid = length(doseGrid),
+    xLevel = matchTolerance(x = x, table = doseGrid),
+    placebo = placebo
+  )
 
+  return(ret)
```

Comment:
avoid explicit return statements, just let the function return implicitly the last value in the function body